### PR TITLE
Refactor Request#prepare!

### PR DESCRIPTION
**Pull request by Ben-M**
I had trouble reading Request#prepare! so I extracted some methods.

I'm not really happy with prepare_body_and_content_type! - perhaps body.to_form should be split into two methods, one of which returns the default content type and one of which returns the body?

### CODE REVIEW

File: lib/hurley/client.rb [hunk 1]
```diff
@@ -214,6 +193,41 @@ def body_receiver
       @body_receiver ||= [nil, BodyReceiver.new]
     end
 
+    def set_authorization_header_from_basic_auth!
+      header[:authorization] = url.basic_auth if url.basic_auth
```
Comment (technoweenie) [inline lib/hurley/client.rb]: Nit picky, but I prefer this:

```ruby
if auth = url.basic_auth
  header[:authorization] = auth
end
```

This only runs `#basic_auth` once.  An alternative is adding a `#basic_auth?` method that checks for the existence of `@user`.

**Comment by technoweenie** [comment]
One alternative to breaking this method up is to use comments to describe each section.  These methods will never be used anywhere else, so they have no real purpose other than code clarity.

**Comment by Ben-M** [comment]
Code clarity is a great purpose!

My experience is that comments tend to gradually drift apart from the actual code, and get separated from the lines they are referring to. Broken comments never cause tests to fail, so people tend not to notice.

What do you think?

File: lib/hurley/client.rb (hunk 1 above)
Comment (Ben-M) [inline lib/hurley/client.rb]: Done. I went with #basic_auth? for readability.

File: test/url_test.rb [hunk 2]
```diff
@@ -407,6 +407,13 @@ def test_basic_auth_user_without_password
       assert_equal "Basic #{Base64.encode64("a b").rstrip}", u.basic_auth
     end
 
+    def test_basic_auth?
+      url_with_user = Url.parse("https://a@example.com")
+      assert_equal url_with_user.basic_auth?, true
+      url_without_user = Url.parse("https://example.com")
+      assert_equal url_without_user.basic_auth?, false
```
Comment (technoweenie) [inline test/url_test.rb]: Expectation first :)  `assert_equal false, url_without_user.basic_auth?`.

You could condense it like this too:

```ruby
assert Url.parse("https://a@example.com").basic_auth?
assert !Url.parse("https://example.com").basic_auth?
```

Comment (Ben-M) [inline test/url_test.rb]: Thanks! Done.

MiniTest is new to me, I appreciate the pointers.

**Comment by technoweenie** [comment]
> My experience is that comments tend to gradually drift apart from the actual code

Same thing happens with variable/method/class names too.  Also, methods with lots of small helper methods can be more confusing to read too.  Though I agree that `Request#prepare!` has crossed that threshold and should be split up.  What if it was something really small like this:

```ruby
    def prepare!
      prepare_basic_auth!

      if body
        prepare_body_form!
      else
        return unless REQUIRED_BODY_VERBS.include?(verb)
      end

      prepare_content_length!
    end
```

Then trust that each helper is doing its thing properly.  We can trust that `set_content_length!` is doing the right thing, instead of asking `content_length_or_chunked_set?`.  That gives you an idea at a glance what it's doing.  If you care about the specifics, you can dig into that method.

Now's the time to move stuff around to.  I'm not sure `RequestOptions#build_form` is in the right place, for example.  Or even `Request#prepare!`.

**Comment by technoweenie** [comment]
I threw my suggestion in #10.

**Pull request closed without merging**